slasher/rpc/detect_update_min_max_span.go

```diff
@@ -0,0 +1,103 @@
+package rpc
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/pkg/errors"
+	ethpb "github.com/prysmaticlabs/prysm/proto/eth/v1alpha1"
+	"github.com/prysmaticlabs/prysm/shared/params"
+)
+
+type detect func(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64
+
+func detectMax(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64 {
+	maxSpan := uint64(spans.MaxSpan)
+	if maxSpan > span {
+		return maxSpan + source
+	}
+	return 0
+}
+
+func detectMin(span uint64, spans *ethpb.MinMaxSpan, source uint64) uint64 {
+	minSpan := uint64(spans.MinSpan)
+	if minSpan < span {
+		return minSpan + source
+	}
+	return 0
+}
+
+func (ss *Server) detectSpan(source, target, validatorIdx uint64, detectionFunc detect) (targetEpoch uint64, span uint64, spanMap *ethpb.EpochSpanMap, err error) {
+	span = target - source + 1
+	if span > params.BeaconConfig().WeakSubjectivityPeriod {
+		return 0, span, nil, fmt.Errorf("%d target - source: %d > weakSubjectivityPeriod",
+			params.BeaconConfig().WeakSubjectivityPeriod,
+			span,
+		)
+	}
+	spanMap, err = ss.SlasherDB.ValidatorSpansMap(validatorIdx)
+	if err != nil {
+		return 0, span, nil, errors.Wrapf(err, "could not retrieve span map for validatorIdx: %d", validatorIdx)
+	}
+	if _, ok := spanMap.EpochSpanMap[source]; ok {
+		return detectionFunc(span, spanMap.EpochSpanMap[source], source), span, spanMap, nil
+	}
+	return 0, span, spanMap, nil
+}
+
+// DetectAndUpdateMaxSpan is used to detect and update the max span of an incoming attestation.
+// logic is following the detection method designed by https://github.com/protolambda
+// from here: https://github.com/protolambda/eth2-surround/blob/master/README.md#min-max-surround
+func (ss *Server) DetectAndUpdateMaxSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (surroundedTargetEpoch uint64, err error) {
```

Comment:
Please don't use named return variables. 

```suggestion
func (ss *Server) DetectAndUpdateMaxSpan(ctx context.Context, source uint64, target uint64, validatorIdx uint64) (uint64, error) {
```

Reply:
removing. thanks